Support public key caching for jwk endpoints (#16937)
crate/crate: docs/admin/auth/methods.rst

If it's provided both in the token and in the response from the JWK endpoint,
both values are compared and in case of a mismatch the token is rejected.

JWT authentication requires retrieving public keys from a `JWK endpoint`_.
CrateDB will cache these public keys for 10 hours, or according to the
``Cache-Control: max-age`` value in the response header from the `JWK endpoint`_,
if provided. This value must be specified in seconds as a positive integer;

> otherwise, it will be ignored.

Reviewer: Given that we don't have default caching I think we must keep information (needs adjustment 10 hours is irrelevant).

Users can forget about adding that header and it will lead to performance decrease as originally reported

Reviewer 2: I don't see why. "The implementation uses Cache-Control headers to cache result" is kind of just doing HTTP right?

Reviewer 2: Okay fair point - added a fixup

If it's provided both in the token and in the response from the JWK endpoint,
both values are compared and in case of a mismatch the token is rejected.

JWT authentication requires retrieving public keys from a `JWK endpoint`_.

> These public keys can be cached by CrateDB by setting ``cache-control: max-age``

Reviewer 3: ```suggestion
These public keys can be cached by CrateDB by setting ``Cache-Control: max-age``
```

While http headers are case insensitive, thats how it is referred to in the specificaiton

Author: 👍

> in the header of the request. The value has to be defined in seconds and must

Reviewer 3: ```suggestion
in the header of the response. The value has to be defined in seconds and must
```

Reviewer: Good catch, I missed that

Author: 👍

If it's provided both in the token and in the response from the JWK endpoint,
both values are compared and in case of a mismatch the token is rejected.

> JWT authentication requires retrieving public keys from a `JWK endpoint`_.
> These public keys will be cached by CrateDB for 10 hours or by retrieving
> ``Cache-Control: max-age`` from the header of the response from the `JWK endpoint`_.
> The value has to be defined in seconds and must be a positive integer otherwise
> it will be ignored.
> 

Reviewer 3: ```suggestion
JWT authentication requires retrieving public keys from a `JWK endpoint`_.
CrateDB will cache these public keys for 10 hours, or according to the
``Cache-Control: max-age`` value in the response header from the `JWK endpoint`_, if provided.
This value must be specified in seconds as a positive integer; otherwise, it will be ignored.

```

Author: 👍

If it's provided both in the token and in the response from the JWK endpoint,
both values are compared and in case of a mismatch the token is rejected.

> 
> JWT authentication requires retrieving public keys from JWK endpoints. These
> public keys can be cached by CrateDB by setting ``cache-control: max-age``
> in the header of the request. If it's provided, once the public keys are
> requested for the first time will be cached for the given time. The value
> has to be defined in seconds and must be a positive integer otherwise it will
> be ignored.

Reviewer: ```suggestion

JWT authentication requires retrieving public keys from a `JWK endpoint`_. 
These public keys can be cached by CrateDB by setting ``cache-control: max-age``
in the header of the request. The value
has to be defined in seconds and must be a positive integer otherwise it will
be ignored.
```

Reviewer: I think we can drop impl-n detail here and in the change entry.

Author: 👍
